### ISSUE TITLE
`keepdim=False` then `broadcast_in_dim` removes explicit alias registers and shared memory can't be reused.

### ISSUE THREAD

**Issue by liqiangxl** [NVIDIA/Fuser#633]
When working on #582 and test softmax, I found that if FusionDefinition use `keepdim=False` then `broadcast_in_dim`, the generated code failed to explicitly use alias registers. There is no influence on performance since the compiler can auto reuse. But this disables shared memory reuse when the register buffers are changed to shared memroy buffers in #582 .
Related to #221 
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguity=[True, True], dtype=DataType.Half, is_cpu=False)
    T1 = fd.ops.cast(T0, dtype=DataType.Float)
    T2 = fd.ops.max(T1, axes=[1], keepdim=False, dtype=DataType.Null)
    T3 = fd.ops.broadcast_in_dim(T2, output_shape=[8192, 1], broadcast_dims=[0])
    T4 = fd.ops.broadcast_in_dim(T3, output_shape=[8192, 10240], broadcast_dims=[0, 1])
    T5 = fd.ops.sub(T1, T4)
    T6 = fd.ops.exp(T5)
    T7 = fd.ops.sum(T6, axes=[1], keepdim=False, dtype=DataType.Null)
    T8 = fd.ops.broadcast_in_dim(T7, output_shape=[8192, 1], broadcast_dims=[0])
    T9 = fd.ops.broadcast_in_dim(T8, output_shape=[8192, 10240], broadcast_dims=[0, 1])
    T10 = fd.ops.div(T6, T9)
    T11 = fd.ops.cast(T10, dtype=DataType.Half)
    fd.add_output(T11)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn((8192, 10240), dtype=torch.float16, device='cuda:0').as_strided((8192, 10240), (10240, 1)),
]
fd.execute(inputs)

```

If `keepdim=True`, alias register is explicitly used and also shared memory.
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguity=[True, True], dtype=DataType.Half, is_cpu=False)
    T1 = fd.ops.cast(T0, dtype=DataType.Float)
    T2 = fd.ops.max(T1, axes=[1], keepdim=True, dtype=DataType.Null)
    # T3 = fd.ops.broadcast_in_dim(T2, output_shape=[8192, 1], broadcast_dims=[0])
    # T4 = fd.ops.broadcast_in_dim(T3, output_shape=[8192, 10240], broadcast_dims=[0, 1])
    T5 = fd.ops.sub(T1, T2)
    T6 = fd.ops.exp(T5)
    T7 = fd.ops.sum(T6, axes=[1], keepdim=True, dtype=DataType.Null)
    # T8 = fd.ops.broadcast_in_dim(T7, output_shape=[8192, 1], broadcast_dims=[0])
    # T9 = fd.ops.broadcast_in_dim(T8, output_shape=[8192, 10240], broadcast_dims=[0, 1])
    T10 = fd.ops.div(T6, T7)
    T11 = fd.ops.cast(T10, dtype=DataType.Half)
    fd.add_output(T11)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn((8192, 10240), dtype=torch.float16, device='cuda:0').as_strided((8192, 10240), (10240, 1)),
]
fd.execute(inputs)

```

Difference in generated code:

![image](https://github.com/NVIDIA/Fuser/assets/116412316/3b17fbb1-ed6a-4b6e-a416-595ce002df5a)

Success shared memory reuse in #582  if `keepdim=True`
![image](https://github.com/NVIDIA/Fuser/assets/116412316/27290142-7bb4-487a-b3c9-e8123f09ca2d)

**Comment by liqiangxl** [comment]
there is an `expand` op in this python script, which is identified as incompatible with memory aliasing. The proposed fix is:
(1) allow `expand` and treat it similar to broadcast, `that is require matched iterdomains for reuse`

If the data type is fp32, the access of the first buffer is vectorized but the 2nd buffer is not vectorized, this also disables the reuse. To fix this, we need to:
(2) Allow non-vectorized tv reuse vectorized tv.
